Full docs of the OPC UA public beta release (#1)
jaspervanderhoek/docs: content/appstore/connectors/opc-ua.md

* Credentials
* Client certificates

> Only one of the options can be in use any any time, which one will be determined by the OPC UA server that you are connecting too. The password from both the user and certificate are automatically encrypted when saving the server configuration using the Encryption module.  

Reviewer: ```suggestion
Only one of the options can be in use any any time. Which one will be determined by the OPC UA server that you are connecting to. The password from both the user and certificate are automatically encrypted when saving the server configuration using the Encryption module.  
```

#### 1.1.3 Dependencies

The OPC UA Client connector has the following dependencies

* Mendix 8.8.1 or higher
* Encryption Module
* An OPC UA server – we assume that you are familiar with OPC UA and your OPC UA server(s), these instructions do not include information on how to set up your servers.

> Recommended; Having an external OPC UA Client tool will make setup of the connection a lot easier. A commonly used client is the one from [Unified Automation](https://www.unified-automation.com/downloads/opc-ua-clients.html), with this (non-Mendix) tool you will be able to browse through the server and find the parameters needed to configure your connection. 

Reviewer: ```suggestion
Recommended; Having an external OPC UA Client tool will make setup of the connection a lot easier. A commonly-used client is the one from [Unified Automation](https://www.unified-automation.com/downloads/opc-ua-clients.html). With this (non-Mendix) tool you will be able to browse through the server and find the parameters needed to configure your connection. 
```

{{% image_container width="300" %}}
![OPC UA Client connector in Project Explorer](attachments/opc-ua/opc-ua-connector-module.png)
{{% /image_container %}}
{{% todo %}}[Update this Image, the folder structure has changed]{{% /todo %}}

When you edit a microflow, you will also see five additional actions in the **Toolbox**.

{{% image_container width="300" %}}
![OPC UA Client connector actions in the microflow toolbox](attachments/opc-ua/opc-ua-toolbox.png)
{{% /image_container %}}

### 2.1 Configuration:  
1. Add the **OpcUaServer_Overview** page to the navigation of the app, either through the **Navigation** settings, or by adding an **Open Page** button to a page which is already in the navigation (for example the home page). 

> 2. Looup the Constant: **UA_ApplicationName** in the **_USE_ME** folder and update these with the information specific to your Client. The application name must be unique to the OPC UA Server, the connector has no requirements to the contents of this constant. If you have a single Mendix application connecting to the server you could choose to leave the default value, if you have multiple Mendix applications connecting to the same server you **must** alter the name to be unique (as per the OPC UA spec).   

Reviewer: ```suggestion
2. Find the constant: **UA_ApplicationName** in the **_USE_ME** folder and update this with the information specific to your Client. The application name must be unique to the OPC UA Server, the connector has no requirements to the contents of this constant. If you have a single Mendix application connecting to the server you could choose to leave the default value, if you have multiple Mendix applications connecting to the same server you **must** alter the name to be unique (as per the OPC UA spec).   
```

*See the OPC UA Specification for more details, this setting is only used by the OPC UA Server.*

> 3. Looup the Constant: **UA_ApplicationURI** in the **_USE_ME** folder and update these with the information specific to your Client. The application URI must be unique to the OPC UA Server, the connector has no requirements to the contents of this constant. If you have a single Mendix application connecting to the server you could choose to leave the default value, if you have multiple Mendix applications connecting to the same server you **must** alter the name to be unique (as per the OPC UA spec).   

Reviewer: ```suggestion
3. Find the constant: **UA_ApplicationURI** in the **_USE_ME** folder and update this with the information specific to your Client. The application URI must be unique to the OPC UA Server, the connector has no requirements to the contents of this constant. If you have a single Mendix application connecting to the server you could choose to leave the default value, if you have multiple Mendix applications connecting to the same server you **must** alter the name to be unique (as per the OPC UA spec).   
```

* URL (String) – the full URL of the OPC UA server (e.g.: opc.tcp://localhost:53530/OPCUA/SimulationServer)
* Username (String) – the username used if the authentication type is `CREDENTIALS`
* Password (String) – the password for the username used if the authentication type is `CREDENTIALS`, only used for data-entry. After saving this field is reset.
* Password_Encrypted (String) – the encrypted password for the username, automatically takes the 'Password' and encrypts the value
* AuthenticationType (Enumeration) – the type of authentication required for this server: `NONE`, `CREDENTIALS`, or `CERTIFICATE`
* CertificatePassword (String) – the certificate password required when using the `CERTIFICATE` type of authentication, only used for data-entry. After saving this field is reset.
* CertificatePassword (String) – the encrypted password for the certificate, automatically takes the CertificatePassword and encrypts the value

You can see an example of how this can be set up in the [OPC UA Client example implementation](#example-implementation) section.

### 3.2 Actions

Once you have set up the server configuration, you can perform the following actions in your microflows.

> #### 3.2.1 **Browse** a List of Nodes

Reviewer: ```suggestion
#### 3.2.1 **Browse** a List of Nodes{#browse}
```

The **Browse** action allows you to browse the nodes within the OPC UA server. The browse function starts at the specified node and browses 'down' and returns the children of the specified node. 

![Parameters for the browse action](attachments/opc-ua/browse-action.png)

* Opc UA server cfg – an object of entity type OpcUaServerCfg containing the configuration of the server to which the request is made
* NodeId – The NodeId from where you want to browse to it's children. When requesting the 'Is Root'=true leave this value empty. Expects the full Node Id as referenced by the OPC UA server. This is generally a combination of the namespace URI and Identifier but can have different variations. You can find this in most OPC UA Clients (including the Unified Automation client) and the Browse function returns this same value for each node. Example: "ns=4;id=3"
* Is root – is used by the tree widget in the example implementation module — if you are not using the tree widget you, the value here is not important
* Return Variable – The output of this action is the JSON string with all information about the requested nodes. This string can be parsed with the Import Mapping activity.

> #### 3.2.2 **Read** the Value of a Node

Reviewer: ```suggestion
#### 3.2.2 **Read** the Value of a Node{#read}
```

The **Read** action allows you to read the current value of a specific node within the OPC UA server. The output of the action is a string formatted value of the Node. While the module supports most OPC UA attribute types the action always returns the value as a string as it was showing in the original message received from the OPC UA Server. Example the decimal value 10.59 will be returned as '10.59'   

![Parameters for the read action](attachments/opc-ua/read-action.png)

* Opc ua server cfg – an object of entity type OpcUaServerCfg containing the configuration of the server to which the request is made
* NodeId – The NodeId of the Node you want to read. Expects the full Node Id as referenced by the OPC UA server. This is generally a combination of the namespace URI and Identifier but can have different variations. You can find this in most OPC UA Clients (including the Unified Automation client) and the Browse function returns this same value for each node. Example: "ns=4;id=3"

{{% alert type="info" %}}
All values are read as strings, you will need to convert them if you need a numeric or date value.
{{% /alert %}}

> #### 3.2.3 **Subscribe** to Updates of Data from a Node

Reviewer: ```suggestion
#### 3.2.3 **Subscribe** to Updates of Data from a Node{#subscribe}
```

**Attention:** Never manually or programatically delete the **Subscription** or **MonitoredItem**, unless the Status is 'Deleted'. Removing the objects prematurely can result in duplicate data or exceptions when the OPC UA Server sends messages for these now removed configurations.  

You can configure if the unsubscripe is permanent (and records are removed) or if the subscription info is kept for restart through the parameters.

![Parameters for the unsubscribe action](attachments/opc-ua/unsubscribe-action.png)

* Opc ua server cfg – an object of entity type OpcUaServerCfg containing the configuration of the server to which the request is made
* Monitored item ID – the ID of the item which is being monitored by the subscription — this is held as the **MonitoredItemID** in the **Subscription** entity
* RestartSubscriptionOnNextReboot – Indicate if the registration entities should be kept by this action. When 'true' the **MonitoredItem** & **Subscription** will be kept in the database and their Status will be changed to 'New'. When 'false' the status of the monitoredItem will become 'Deleted' and the module will automatically remove the entities from the database. 

> #### 3.2.5 **Write** Data to a Node

Reviewer: ```suggestion
#### 3.2.5 **Write** Data to a Node{#write}
```

**Test Messaging:**  
* **NodeId** – The NodeId of the Node you want to use. Expects the full Node Id as referenced by the OPC UA server. This is generally a combination of the namespace URI and Identifier but can have different variations. You can find this in most OPC UA Clients (including the Unified Automation client) and the Browse function returns this same value for each node. Example: "ns=4;id=3"
* **Value to write** – the value to write to a node when the **Write** button is clicked. Leave blank if you want to execute any of the other actions.
* **Result** – the resulting JSON string from performing any of the four test actions. This shows the full response from the OPC UA Server, the connector might at relevant details in case of an error. 

> * **Read** executes the Read action {{% todo %}}[Add a link to the read-action chapter]{{% /todo %}}

Reviewer: ```suggestion
* **Read** executes the [Read action](#read)
```

> * **Write** executes the Write action {{% todo %}}[Add a link to the Write-action chapter]{{% /todo %}}

Reviewer: ```suggestion
* **Write** executes the [Write action](#write)
```

> * **Browse** executes the Browse action {{% todo %}}[Add a link to the Browse-action chapter]{{% /todo %}}

Reviewer: ```suggestion
* **Browse** executes the [Browse action](#browse)
```

> * **Subscribe Simple** executes the Subscribe action {{% todo %}}[Add a link to the Subscribe-action chapter]{{% /todo %}}. Process response with microflow: SubscriptionIncomingData_Process_TEST. Check the log or put a breakpoint on this microflow to checkout the response messages.

Reviewer: ```suggestion
* **Subscribe Simple** executes the [Subscribe action](#subscribe). Process response with microflow: SubscriptionIncomingData_Process_TEST. Check the log or put a breakpoint on this microflow to checkout the response messages.
```

> * **Subscribe w/ Msg**  executes the Subscribe action {{% todo %}}[Add a link to the Subscribe-action chapter]{{% /todo %}}. Process response with microflow: SubscriptionIncomingData_Process_wMessage_TEST. Check the log or put a breakpoint on this microflow to checkout the response messages.

Reviewer: ```suggestion
* **Subscribe w/ Msg**  executes the [Subscribe action](#subscribe). Process response with microflow: SubscriptionIncomingData_Process_wMessage_TEST. Check the log or put a breakpoint on this microflow to checkout the response messages.
```
